shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/EpochBoundary.hs

```diff
@@ -219,46 +240,222 @@ instance
+    !(Map.Map Ptr (Credential 'Staking (Crypto era))) ->
+    -- We do make SDP lazy in the map of active credentials, so it is not computed at the
+    -- at the Epoch Boundary, but only at the first Tick. After that it will already be in normal form.
+    (Map.Map (Credential 'Staking (Crypto era)) (KeyHash 'StakePool (Crypto era))) ->
+    !ans ->
+    StakeDistrPulser era m ans
+
+deriving instance (Eq (UTxO era), Eq ans) => Eq (StakeDistrPulser era m ans)
+
+deriving instance (Show (UTxO era), Show ans) => Show (StakeDistrPulser era m ans)
+
+instance (Era era, NoThunks (UTxO era), c ~ (Crypto era)) => NoThunks (StakeDistrPulser era m (SnapShot c)) where
+  showTypeOf _ = "StakeDistrPulser"
+  wNoThunks ctxt (SDP n u ptrs activedelegs ans) =
+    allNoThunks
+      [ noThunks ctxt n,
+        noThunks ctxt u,
+        noThunks ctxt ptrs,
+        noThunks ctxt activedelegs,
+        noThunks ctxt ans
+      ]
+
+instance (Era era, NFData (UTxO era), c ~ (Crypto era)) => NFData (StakeDistrPulser era m (SnapShot c)) where
+  rnf (SDP n u ptrs active ans) = seq (rnf n) (seq (rnf u) (seq (rnf ptrs) (seq (rnf active) (rnf ans))))
+
+instance (Typeable m, Era era, c ~ Crypto era, ToCBOR (UTxO era)) => ToCBOR (StakeDistrPulser era m (SnapShot c)) where
+  toCBOR (SDP n utxo ptrs active ans) =
+    encode (Rec SDP !> To n !> To utxo !> mapEncode ptrs !> mapEncode active !> To ans)
+
+instance (Typeable m, Era era, c ~ Crypto era, FromCBOR (UTxO era)) => FromCBOR (StakeDistrPulser era m (SnapShot c)) where
+  fromCBOR =
+    decode
+      ( RecD SDP <! From <! From <! mapDecode <! mapDecode <! From
+      )
+
+-- ====================================================================
+
+-- NOW make it an instance of Pulsable, for (StakeDistrPulser era)
+
+instance
+  ( Era era,
+    c ~ Crypto era
+  ) =>
+  Pulsable (StakeDistrPulser era)
+  where
+  done (SDP _n (UTxO u) _ptrs _active _ans) = Map.null u
+  current (SDP _n _utxo _ptrs _active ans) = ans
+  pulseM ll@(SDP _n (UTxO u) _ptrs _active _ans) | Map.null u = pure ll
+  pulseM (SDP n (UTxO u) ptrs active (SnapShot (Stake b) ds ps)) = do
+    let (!steps, !smaller_u) = Map.splitAt n u
+        new = SnapShot (Stake (aggregateUtxoCoinByActiveCredential @era ptrs active (UTxO steps) b)) ds ps
+    pure (SDP n (UTxO smaller_u) ptrs active new)
+  completeM (SDP _n utxo ptrs active (SnapShot (Stake b) ds ps)) =
+    let stakeRelation = aggregateUtxoCoinByActiveCredential @era ptrs active utxo b
+     in pure $ SnapShot (Stake stakeRelation) ds ps
+
+-- | Sum up all the Coin for each active staking Credential
+aggregateUtxoCoinByActiveCredential ::
+  forall era.
+  Era era =>
+  Map.Map Ptr (Credential 'Staking (Crypto era)) ->
+  Map.Map (Credential 'Staking (Crypto era)) (KeyHash 'StakePool (Crypto era)) ->
+  UTxO era ->
+  Map.Map (Credential 'Staking (Crypto era)) Coin ->
+  Map.Map (Credential 'Staking (Crypto era)) Coin
+aggregateUtxoCoinByActiveCredential ptrs active (UTxO u) initial =
```

Comment:
Do you expect this to be much faster than using `aggregateCoinByCredential` and then filtering to the active delegations? It otherwise duplicates pretty much all of that function.

Reply:
I guess that depends upon number of Inactive credentials vs the number of active credentials.  This also makes the pulser smaller, since the answer now only stores results that have active credentials, so there is a space as well as time dimension.  Some profiling is in order. Except profiing is usually done with artifical UTxO and Credential maps, so not sure if it could tell us much.

Reply:
Also we only need one or the other. They both have the same inputs and compute the same outputs. I added it because of the size of the intermediate pulsers concern. But we could use it everywhere.

Reply:
Eventually it should replace aggregateCoinByCredential. It has the same type and computes the same thing. It has the same asymptotic complexity but the constant should be smaller.